### ISSUE TITLE
pgsqld failed in pacemaker with unknown error

### ISSUE THREAD

**Issue by ARogovsky** [dalibo/PAF#11]
Hi
I try run cluster with PAF and received this error:
`pgsqld_start_0 on srvc 'unknown error' (1): call=541, status=complete, exit-reason='none', last-rc-change='Mon Apr  4 12:33:16 2016', queued=0ms, exec=5128ms`
Same errors on srva and srvb

Postgresql in not running and no running attempts in log files.
How I can debug PAF and show reason of error?

**Comment by ioguix** [comment]
Hello,

What is your Linux distribution (exactly)?

Could you please give us information about the PostgreSQL and Pacemaker setup?

Regards,

**Comment by ARogovsky** [comment]
```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.1 (jessie)
Release:	8.1
Codename:	jessie

# crm configure show
node 1084754433: a.server \
node 1084754434: b.server \
node 1084754435: c.server \
primitive pgsql-master-ip IPaddr2 \
	params ip=192.168.10.200 nic=peervpn0 \
	op start interval=0s on-fail=restart timeout=60s \
	op monitor interval=10s on-fail=restart timeout=60s \
	op stop interval=0s on-fail=block timeout=60s \
	meta target-role=Started
primitive pgsqld pgsqlms \
	params bindir="/usr/lib/postgresql/9.4/bin" pgdata="/var/lib/postgresql/9.4/main" \
	op start interval=0 on-fail=restart timeout=160s \
	op monitor interval=40s on-fail=restart timeout=160s \
	op promote interval=0s on-fail=restart timeout=160s \
	op demote interval=0s on-fail=stop timeout=160s \
	op stop interval=0 on-fail=block timeout=160s \
	op notify interval=0s timeout=160s
group master pgsql-master-ip
ms msPostgresql pgsqld \
	meta master-max=1 master-node-max=1 clone-max=3 clone-node-max=1 target-role=Started notify=true is-managed=true
colocation set_ip inf: master msPostgresql:Master
order ip_down 0: msPostgresql:demote master:stop symmetrical=false
order ip_up 0: msPostgresql:promote master:start symmetrical=false

```

**Comment by ioguix** [comment]
Hello,

Could you please post your `/etc/corosync/corosync.conf`?

I believe adding `debug: on`in your `logging` section in `/etc/corosync/corosync.conf` should be enough. Eg.:

```
[...]
logging {
	to_logfile: yes
	logfile: /var/log/corosync/corosync.log
	to_syslog: yes
	syslog_facility: daemon
	debug: on
	timestamp: on
	logger_subsys {
		subsys: QUORUM
		debug: off
	}
}
```

**Comment by ARogovsky** [comment]
```
# cat /etc/corosync/corosync.conf
# Please read the openais.conf.5 manual page

totem {
	version: 2

crypto_compat 2.0

        crypto_cipher: none
        crypto_hash: none

	# How long before declaring a token lost (ms)
	token: 2000

	# How many token retransmits before forming a new configuration
	token_retransmits_before_loss_const: 10

	# How long to wait for join messages in the membership protocol (ms)
	join: 100

	# How long to wait for consensus to be achieved before starting a new round of membership configuration (ms)
	consensus: 2500

	# Turn off the virtual synchrony filter
	vsftype: none

	# Number of messages that may be sent by one processor on receipt of the token
	max_messages: 20

	# Limit generated nodeids to 31-bits (positive signed integers)
	clear_node_high_bit: yes

	# Disable encryption
 	secauth: off

	# How many threads to use for encryption/decryption
 	threads: 0

	# Optionally assign a fixed node id (integer)
	# nodeid: 1234

	# This specifies the mode of redundant ring, which may be none, active, or passive.
 	rrp_mode: none

 	interface {
		# The following values need to be set based on your environment 
		ringnumber: 0
		bindnetaddr: 192.168.10.0
		mcastaddr: 239.255.1.1
		mcastport: 5406
	}


}

amf {
	mode: disabled
}

service {
 	# Load the Pacemaker Cluster Resource Manager
 	ver:       0
 	name:      pacemaker
}

aisexec {
        user:   root
        group:  root
}

logging {
        fileline: off
        to_stderr: yes
        to_logfile: no
        to_syslog: yes
	syslog_facility: daemon
        debug: off
        timestamp: on
        logger_subsys {
                subsys: AMF
                debug: off
                tags: enter|leave|trace1|trace2|trace3|trace4|trace6
        }
}

quorum {
    provider: corosync_votequorum
    expected_votes: 2
}

```

**Comment by ioguix** [comment]
So did you try to turn on debug in the corosync.conf file as i advice you earlier today?

**Comment by ARogovsky** [comment]
Allrigjt
Log file is big, I was upload it here: http://www.filedropper.com/logtxt
This most important strings:
```
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: pgsql_monitor: monitor is a probe
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: _runas: launching as "postgres" command "/usr/lib/postgresql/9.4/bin/pg_isready -h /tmp -p 5432"
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: pgsql_monitor: instance "pgsqld" is not listening
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: _runas: launching as "postgres" command "/usr/lib/postgresql/9.4/bin/pg_ctl -D /var/lib/postgresql/9.4/main status"
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: _confirm_stopped: no postmaster process found for instance "pgsqld"
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: _controldata: instance "pgsqld" state is "shut down in recovery"
pgsqlms(pgsqld)[1886]:  2016/04/06_08:11:46  DEBUG: _confirm_stopped: instance "pgsqld" controldata indicates that the instance was propertly shut down
...
Apr 06 08:11:51 [1878] a.geocluster.e-autopay.com       lrmd:    debug: operation_finished:     pgsqld_start_0:2017 - exited with rc=1
Apr 06 08:11:51 [1878] a.geocluster.e-autopay.com       lrmd:   notice: operation_finished:     pgsqld_start_0:2017:stderr [ pg_ctl: could not start server ]
Apr 06 08:11:51 [1878] a.geocluster.e-autopay.com       lrmd:   notice: operation_finished:     pgsqld_start_0:2017:stderr [ Examine the log output. ]
Apr 06 08:11:51 [1878] a.geocluster.e-autopay.com       lrmd:    debug: operation_finished:     pgsqld_start_0:2017:stdout [ /tmp:5432 - no response ]
Apr 06 08:11:51 [1878] a.geocluster.e-autopay.com       lrmd:    debug: operation_finished:     pgsqld_start_0:2017:stdout [ pg_ctl: no server running ]
Apr 06 08:11:51 [1878] a.geocluster.e-autopay.com       lrmd:    debug: operation_finished:     pgsqld_start_0:2017:stdout [ waiting for server to start....postgres cannot access the 
server configuration file "/var/lib/postgresql/9.4/main/postgresql.conf": No such file or directory ]
```
My configs in the /etc/postgresql/9.4/main
How I can configure PAF for sent path to configs directory?

**Comment by ioguix** [comment]
Hello,

Currently, the only way to have PAF working under Debian is to not use the pg_wrappers and follow the PostgreSQL default setup: having configuration files in the PGDATA.

We will make a patch to add a parameter to the resource agent for a better integration with Debian.

We'll close this issue after the patch will be published.

**Comment by ARogovsky** [comment]
Allright, I try create symlnks from pgdata to my etc

**Comment by ARogovsky** [comment]
Now, I can run postgres from command line, same as used in PAF:

```
 root@a:~# sudo -u postgres /usr/lib/postgresql/9.4/bin/pg_ctl -D /var/lib/postgresql/9.4/main -w start
could not change directory to "/root": Отказано в доступе
waiting for server to start....2016-04-07 11:34:49 MSK LOG:  database system was shut down in recovery at 2016-04-07 11:30:06 MSK
2016-04-07 11:34:49 MSK LOG:  database system was not properly shut down; automatic recovery in progress
2016-04-07 11:34:49 MSK LOG:  consistent recovery state reached at 37B/717CE108
2016-04-07 11:34:49 MSK LOG:  record with zero length at 37B/717CE108
2016-04-07 11:34:49 MSK LOG:  redo is not required
2016-04-07 11:34:49 MSK LOG:  MultiXact member wraparound protections are now enabled
2016-04-07 11:34:49 MSK LOG:  database system is ready to accept connections
2016-04-07 11:34:49 MSK LOG:  autovacuum launcher started
 done
server started
root@a:~# sudo -u postgres /usr/lib/postgresql/9.4/bin/pg_ctl -D /var/lib/postgresql/9.4/main stop
could not change directory to "/root": Отказано в доступе
waiting for server to shut down....2016-04-07 11:35:03 MSK LOG:  received smart shutdown request
2016-04-07 11:35:03 MSK LOG:  autovacuum launcher shutting down
2016-04-07 11:35:03 MSK LOG:  shutting down
2016-04-07 11:35:03 MSK LOG:  database system is shut down
 done
server stopped
root@a:~# 
```

```
So, it should work in Pacemaker. But this is permission issue:
pgsqlms(pgsqld)[1827]:  2016/04/07_11:37:35  ERROR: pgsql_start: instance "pgsqld" failed to start (rc: 1)
Apr 07 11:37:35 [1681] a.server       lrmd:    debug: operation_finished:     pgsqld_start_0:1827 - exited with rc=1
Apr 07 11:37:35 [1681] a.server       lrmd:   notice: operation_finished:     pgsqld_start_0:1827:stderr [ pg_ctl: could not start server ]
Apr 07 11:37:35 [1681] a.server       lrmd:   notice: operation_finished:     pgsqld_start_0:1827:stderr [ Examine the log output. ]
Apr 07 11:37:35 [1681] a.server       lrmd:    debug: operation_finished:     pgsqld_start_0:1827:stdout [ /tmp:5432 - no response ]
Apr 07 11:37:35 [1681] a.server       lrmd:    debug: operation_finished:     pgsqld_start_0:1827:stdout [ pg_ctl: no server running ]
Apr 07 11:37:35 [1681] a.server       lrmd:    debug: operation_finished:     pgsqld_start_0:1827:stdout [ waiting for server to start....2016-04-07 11:37:30 MSK FATAL:  could not access private key file "/etc/ssl/private/ssl-cert-snakeoil.key": Permission denied ]
Apr 07 11:37:35 [1681] a.server       lrmd:    debug: operation_finished:     pgsqld_start_0:1827:stdout [ .... stopped waiting ]
Apr 07 11:37:35 [1681] a.server       lrmd:     info: log_finished:   finished - rsc:pgsqld action:start call_id:18 pid:1827 exit-c

```
Why appears permission error? User postgres able to read key file and server start correctly from command line. I have no ideas...

**Comment by ARogovsky** [comment]
I was added users postres and hacluster to read keyfile, but bug is not fixed...

**Comment by ioguix** [comment]
PAF does not start PostgreSQL using sudo but by doing a setiud from root.

Considering my Debian 8 environment, user postgres should be in the `ssl-cert` group to be able to access this private certificat:

```
# id postgres
uid=124(postgres) gid=133(postgres) groupes=133(postgres),109(ssl-cert)

# ls -ld /etc/ssl/private/
drwx--x--- 2 root ssl-cert 4096 nov.   9 11:24 /etc/ssl/private/

# ls -l /etc/ssl/private/ssl-cert-snakeoil.key 
-rw-r----- 1 root ssl-cert 1704 nov.   9 11:24 /etc/ssl/private/ssl-cert-snakeoil.key
```

How did you install PostgreSQL on your Debian system ?
Regards,

**Comment by ARogovsky** [comment]
Yes, I was added userpostgres to ssl-cert group. And hacluster user too. This is not help.
I was instal PostgreSQL from apt-get.

**Comment by ARogovsky** [comment]
Allright, this looks as Corosync/Pacemaker issue.
I was copy cert and key files to pgdata and set owner postgres, group haclien, chmod 640.
Your agent works fine!
For test reason I was create file owner root, group postgres, permission 640. In debug log I was show permission problem.
So, lrmd switch owner, but forget switch group and leave it haclient.

Thanks for help. I will test your agent soon.

**Comment by ioguix** [comment]
Hi,

Thank you for your feedback.

I keep this issue open to keep track of the fix concerning the postgresql.conf file.

About the rights issue, I don't understand what you mean by:

> So, lrmd switch owner, but forget switch group and leave it haclient.

Regards,

**Comment by ARogovsky** [comment]
I meant, when Postgresql started as postgres:haclient instead postgres:postgres.

**Comment by ioguix** [comment]
Please, keep this issue open, we have to fix the postgresql.conf location issue and investigate about this rights trouble to make sure it does not come from our side.

Regards,

**Comment by ARogovsky** [comment]
Allright, will wait.

**Comment by ioguix** [comment]
Hello,

So I just committed two different patches about:
  * supporting custom argument on postgres startup (to give config_file path)
  * fix your rights issue by setting the postgres (or the given user in param `system_user`) groups before running any command with its rights.

Good news! That means you don't have to mess with user or files rights anymore concerning the `ssl-cert-snakeoil.key` file!

Moreover, I created a quick start page that might help you and others to install and setup PAF under Debian 8: http://dalibo.github.io/PAF/Quick_Start-Debian-8.html

Last thing, I plan to publish Debian packages for next release. 

Thank you for your valuable tests and reports ARogovsky!
